core/src/replay_stage.rs

```diff
@@ -149,6 +149,62 @@ struct SkippedSlotsInfo {
     last_skipped_slot: u64,
 }
 
+struct PartitionInfo {
+    partition_exists: bool,
+    partition_start_time: Option<Instant>,
+}
+
+impl PartitionInfo {
+    fn new() -> Self {
+        Self {
+            partition_exists: false,
+            partition_start_time: None,
+        }
+    }
+
+    fn update(
+        &mut self,
+        partition_detected: bool,
+        heaviest_bankslot: Slot,
+        last_voted_slot: Slot,
+        reset_bank_slot: Slot,
```

Comment:
while we're at it could we also add the switch fork decision / heaviest fork failure. particularly interested in the `FailedSwitchThreshold` and `FailedThreshold` as it could be useful for debugging.

Reply:
this is to augment the `replay_stage-partition-start` datapoint with an extra "cause of failure" information? I.e. logging why we can't switch/vote on the heaviest fork? I'm guessing this would be propagated from  the `SelectVoteAndResetForkResult` returned by `select_vote_and_reset_forks()`, sampled below

```
 if is_locked_out {
                failure_reasons.push(HeaviestForkFailures::LockedOut(bank.slot()));
            }
            if !vote_threshold {
                failure_reasons.push(HeaviestForkFailures::FailedThreshold(bank.slot()));
            }
            if !propagation_confirmed {
                failure_reasons.push(HeaviestForkFailures::NoPropagatedConfirmation(bank.slot()));
            }
```

This one's interesting b/c there could be multiple reasons that change as the partition continues. For instance we could be both:

1. Locked out
2. FailedSwitchThreshold

so this field would have to be a list

Reply:
Took a stab at piping switch fork decision and heaviest fork failure into the update. Currently just takes the first 2 failures (if they exist) and populates them into static fields. It's also storing these as `Option<T>` formatted as `String`.

It's pretty messy and not exactly DB friendly, but maybe a good starting point for a discussion on how to do this right :)